Smt/Data/Sexp.lean

```diff
@@ -23,21 +26,29 @@ def serializeMany (ss : List Sexp) : String :=
 instance : ToString Sexp :=
   ⟨serialize⟩
 
-partial def parse (s : String) : Except String (List Sexp) :=
-  let tks := tokenize #[] s.toSubstring
+partial def parse (s : String) : Except String (List Sexp) := do
+  let tks ← tokenize #[] s.toSubstring
   parseMany #[] tks.toList |>.map Prod.fst |>.map Array.toList
 where
-  tokenize (stk : Array Substring) (s : Substring) : Array Substring :=
-    if s.isEmpty then stk
-    else
-      let c := s.front
-      if c == ')' || c == '(' then
-        tokenize (stk.push <| s.take 1) (s.drop 1)
-      else if c.isWhitespace then tokenize stk (s.drop 1)
-      else
-        let tk := s.takeWhile fun c => !c.isWhitespace && c != '(' && c != ')'
-        if tk.bsize > 0 then tokenize (stk.push tk) (s.extract ⟨tk.bsize⟩ ⟨s.bsize⟩)
-        else unreachable!
+  tokenize (stk : Array Substring) (s : Substring) : Except String (Array Substring) := do
+    if s.isEmpty then
+      return stk
+    let c := s.front
+    if c == '"' then
+      let s1 := s.drop 1 |>.takeWhile (· ≠ '"')
+      if s1.stopPos = s.stopPos then
+        throw "ending \" missing"
+      let s1 := ⟨s.str, s.startPos, s.next s1.stopPos⟩
+      let s2 := ⟨s.str, s1.stopPos, s.stopPos⟩
+      return (← tokenize (stk.push s1) s2)
```

Comment:
I am not sure if the standard S-expr supports string literals, but I think they're popular enough to support in your parser. I made this change to parse errors generated by the solvers. We also to parse quoted symbols, but that's specific to SMT-LIB. We can either refactor the parser to only parse SMT-LIB S-exprs or make it extensible and specialize it for our specific use case. I will leave the decision to you 🙂. I also tried to get the DSL to support string literals, but couldn't get it to fully work...

Reply:
Let's support them! There is no "S-exp standard", we should add all the features we need.